text/0000-default-field-values.md

```diff
@@ -0,0 +1,1979 @@
+    Ingredient::Tomato { taste: Delicious, color: Color::Green, },
+    Ingredient::Onion { .. },
+];
+```
+
+Note that `enum` variants have public fields and in today's Rust,
+this cannot be controlled with visibility modifiers on variants.
+
+Furthermore, when `#[non_exhaustive]` is specified directly on an `enum`,
+it has no interaction with the defaults values and the ability to construct
+variants of said enum. However, as specified by [RFC 2008], `#[non_exhaustive]`
+is permitted on variants. When that occurs, the behaviour is the same as if
+it had been attached to a `struct` with the same fields and field visibility.
+
+### Interaction with `#[default]`
+
+[default]: https://github.com/rust-lang/rfcs/pull/3107
+
+It is possible today to specify a [`#[default]` variant][default] in an enum so that it
+can be `#[derive(Default)]`. A variant marked with `#[default]` will use
+defaulted fields when present.
+
+```rust
+#[derive(Default)]
+enum Ingredient {
+    Tomato {
+        color: Color = Color::Red,
+        taste: TasteQuality,
+    },
+    Onion {
+        color: Color = Color::Yellow,
+    },
+    #[default]
+    Lettuce {
+        color: Color = Color::Green,
+    },
+}
+```
+
+Now the compiler does know that `Ingredient::Lettuce` should be considered
+the default and will accordingly generate an appropriate implementation of
+`Default for Ingredient` (19):
+
+```rust
+impl Default for Ingredient {
+    fn default() -> Self {
+        Ingredient::Lettuce {
+            color: Color::Green,
+        }
+    }
+}
+```
+
+## Defaults on tuple `struct`s and tuple `enum` variants
+
+Default values are only allowed on named fields. There is no syntax provided for
+tuple types like `struct S(i32)` or `enum E { V(i32), }`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Field default values
+
+### Grammar
+
+Let the grammar of record fields in `struct`s and `enum` variants be defined
+like so (in the `.lyg` notation):
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type;
+```
+
+Then, `RecordField` is changed into:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type { "=" def:Expr }?;
+```
+
+Further, given the following partial definition for the expression grammar:
+
+```rust
+Expr = attrs:OuterAttr* kind:ExprKind;
+ExprKind =
+  | ...
+  | Struct:{ path:Path "{" attrs:InnerAttr* fields:StructExprFieldsAndBase "}" }
+  ;
+
+StructExprFieldsAndBase =
+  | Fields:{ fields:StructExprField* % "," ","? }
+  | Base:{ ".." base:Expr }
+  | FieldsAndBase:{ fields:StructExprField+ % "," "," ".." base:Expr }
+  ;
+StructExprField = attrs:OuterAttr* kind:StructExprFieldKind;
+StructExprFieldKind =
+  | Shorthand:IDENT
+  | Explicit:{ field:FieldName ":" expr:Expr }
+  ;
+```
+
+the rule `StructExprFieldsAndBase` is extended with:
+
+```rust
+StructExprFieldsAndBase =| FieldsAndDefault:{ fields:StructExprField+ % "," "," ".." };
```

Comment:
Why not just frame it as updating `Base` and `FieldsAndBase` so that `base` is optional?